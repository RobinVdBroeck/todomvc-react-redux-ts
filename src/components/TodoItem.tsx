import classnames from "classnames";
import * as React from "react";
import { ITodo } from "../interfaces/ITodo";
import { TodoTextInput } from "./TodoTextInput";

export interface IProps {
  todo: ITodo;
  editTodo(id: number, text: string): void;
  deleteTodo(id: number): void;
  completeTodo(id: number): void;
}

interface IState {
  editing: boolean;
}

export class TodoItem extends React.Component<IProps, IState> {
  private readonly handleSave = this.handleSaveFactory(this.props.todo.id);

  constructor(props: IProps) {
    super(props);
    this.state = { editing: false };
  }

  public render() {
    const { todo } = this.props;

    let element;
    if (this.state.editing) {
      element = (
        <TodoTextInput
          text={todo.text}
          editing={this.state.editing}
          onSave={this.handleSave}
        />
      );
    } else {
      element = (
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            checked={todo.completed}
            onChange={this.onChange}
          />
          <label onDoubleClick={this.handleDoubleClick}>
            {todo.text}
          </label>
          <button className="destroy" onClick={this.onClick} />
        </div>
      );
    }

    return (
      <li
        className={classnames({
          completed: todo.completed,
          editing: this.state.editing
        })}
      >
        {element}
      </li>
    );
  }

  private handleDoubleClick = () => {
    this.setState({ editing: true });
  };

  private onChange = () => this.props.completeTodo(this.props.todo.id);
  private onClick = () => this.props.deleteTodo(this.props.todo.id);

  private handleSaveFactory(id: number) {
    return (text: string) => {
      if (text.length === 0) {
        this.props.deleteTodo(id);
      } else {
        this.props.editTodo(id, text);
      }
      this.setState({ editing: false });
    };
  }
}
