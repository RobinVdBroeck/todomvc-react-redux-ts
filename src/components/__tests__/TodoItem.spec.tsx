import React from "react";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TodoItem } from "../TodoItem";

const defaultProps = {
  todo: {
    id: 0,
    text: "Use Redux",
    completed: false,
  },
  editTodo: jest.fn(),
  deleteTodo: jest.fn(),
  completeTodo: jest.fn(),
};

test("initial render", () => {
  const { asFragment } = render(<TodoItem {...defaultProps} />);

  expect(asFragment()).toMatchSnapshot();
});

test("marking item as completed should call completetodo", () => {
  render(<TodoItem {...defaultProps} />);

  const input = screen.getByRole("checkbox");
  userEvent.click(input);

  expect(defaultProps.completeTodo).toHaveBeenCalledWith(0);
});

test("removing item should call deleteTodo", () => {
  render(<TodoItem {...defaultProps} />);

  const button = screen.getByRole("button");
  userEvent.click(button);

  expect(defaultProps.deleteTodo).toHaveBeenCalledWith(0);
});

test("double click should put component in edit state", () => {
  render(<TodoItem {...defaultProps} />);

  const listItem = screen.getByRole("listitem");
  const label = within(listItem).getByText("Use Redux");
  userEvent.dblClick(label);

  expect(listItem).toHaveClass("editing");
});

test("pressing enter while editing should call editTodo", () => {
  render(<TodoItem {...defaultProps} />);

  const listItem = screen.getByRole("listitem");
  const label = within(listItem).getByText("Use Redux");
  userEvent.dblClick(label);

  const input = screen.getByRole("textbox");
  userEvent.type(input, " twice{enter}");

  expect(defaultProps.editTodo).toBeCalledWith(0, "Use Redux twice");
});

test("emptying input should call deleteTodo", () => {
  render(<TodoItem {...defaultProps} />);

  const listItem = screen.getByRole("listitem");
  const label = within(listItem).getByText("Use Redux");
  userEvent.dblClick(label);

  const input = screen.getByRole("textbox");
  userEvent.clear(input);
  userEvent.type(input, "{enter}");

  expect(defaultProps.deleteTodo).toHaveBeenCalledWith(0);
});

test("pressing enter should exit component from edit state", () => {
  render(<TodoItem {...defaultProps} />);

  const listItem = screen.getByRole("listitem");
  const label = within(listItem).getByText("Use Redux");
  userEvent.dblClick(label);

  const input = screen.getByRole("textbox");
  userEvent.type(input, "{enter}");

  expect(listItem).not.toHaveClass("editing");
});
