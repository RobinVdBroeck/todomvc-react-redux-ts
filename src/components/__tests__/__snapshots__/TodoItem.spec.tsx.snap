// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render 1`] = `
<DocumentFragment>
  <li
    class=""
  >
    <div
      class="view"
    >
      <input
        class="toggle"
        type="checkbox"
      />
      <label>
        Use Redux
      </label>
      <button
        class="destroy"
      />
    </div>
  </li>
</DocumentFragment>
`;
